packages/caliper-core/test/worker/rate-control/recordRate.js

```diff
@@ -63,42 +73,209 @@ describe('RecordRate controller', () => {
+            const testMessage = new TestMessage('test', [], msgContent);
+            const rateController = RecordRate.createRateController(testMessage, stubStatsCollector, 0);
+            const mockRate = MockRate.createRateController();
+            mockRate.reset();
+            mockRate.isApplyRateControlCalled().should.equal(false);
+            await rateController.applyRateControl();
+            mockRate.isApplyRateControlCalled().should.equal(true);
+        });
+
+        it('should replace path template placeholders for various worker and round indices', () => {
+            const testCases = [
+                { testRound: 0, workerIndex: 0, expectedPath: '../tx_records_client0_round0.txt' },
+                { testRound: 1, workerIndex: 2, expectedPath: '../tx_records_client2_round1.txt' },
+                { testRound: 5, workerIndex: 3, expectedPath: '../tx_records_client3_round5.txt' },
+                { testRound: 10, workerIndex: 7, expectedPath: '../tx_records_client7_round10.txt' },
+            ];
+
+            testCases.forEach(({ testRound, workerIndex, expectedPath }) => {
+                const content = JSON.parse(JSON.stringify(msgContent));
+                content.testRound = testRound;
+                const testMessage = new TestMessage('test', [], content);
+                const controller = RecordRate.createRateController(testMessage, stubStatsCollector, workerIndex);
+                controller.pathTemplate.should.equal(util.resolvePath(expectedPath));
+            });
+        });
+
+        it('should throw an error if the rate controller to record is unknown', async () => {
+            msgContent.rateControl.opts.rateController.type = 'nonexistent-rate';
+            msgContent.rateControl.opts.logEnd = true;
+            const testMessage = new TestMessage('test', [], msgContent);
+
+            (() => {
+                RecordRate.createRateController(testMessage, stubStatsCollector, 0);
+            }).should.throw(/Module "nonexistent-rate" could not be loaded/);
+        });
+
+        it('should throw an error if rateController is undefined', () => {
+            msgContent.rateControl.opts.rateController = undefined;
+            const testMessage = new TestMessage('test', [], msgContent);
+
+            (() => {
+                RecordRate.createRateController(testMessage, stubStatsCollector, 0);
+            }).should.throw('The rate controller to record is undefined');
+        });
+    });
+
+    describe('When Creating a RecordRate Controller', () => {
+        it('should initialize records array if the number of transactions is provided', () => {
+            const testMessage = new TestMessage('test', [], msgContent);
+            sinon.stub(testMessage, 'getNumberOfTxs').returns(5);
+
+            const controller = RecordRate.createRateController(testMessage, stubStatsCollector, 0);
+
+            controller.records.should.be.an('array').that.has.lengthOf(5);
```

Comment:
this test is not correct, as stated in a previous review, the value here should be one more than the value returned by getNumberOfTxs because we can and probably will have an entry for 0 transactions submitted therefore if we have 5 transactions submitted then we would have 0-5 possible records. This will then show there is a bug in the code which needs to be fixed.

Reply:
Correct, we want to pre-allocate an array that can cope with 0 to T elements where T is the number of transactions, therefore we need the array allocate T+1 elements. try it for yourself:
```
let numTx = 5;
let x = new Array(numTx);
console.log(x.length) // returns 5
x[numTx]=1;
console.log(x.length) // returns 6
```

Reply:
oh yeah it is obvious now, i was focussed on guarding the existing implementation, that i lost track of how it was meant to be used